File: Python/July_problems/problem2196_15th_july_2024.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


def create_binary_tree(descriptions):
    nodes = {}
    has_parent = {}

    for desc in descriptions:
        parent_val, child_val, is_left = desc
        if parent_val not in nodes:
            nodes[parent_val] = TreeNode(parent_val)
        if child_val not in nodes:
            nodes[child_val] = TreeNode(child_val)
        if is_left:
            nodes[parent_val].left = nodes[child_val]
        else:
            nodes[parent_val].right = nodes[child_val]
        has_parent[child_val] = True

    root = None
    for node_val in nodes:
        if node_val not in has_parent:
            root = nodes[node_val]
            break

    return root

File: CPP/July_Problems/problem2196_15th_july_2024.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

TreeNode *createBinaryTree(vector<vector<int>> &descriptions)
{
    unordered_map<int, TreeNode *> nodes;
    unordered_map<int, bool> hasParent;

    for (const auto &desc : descriptions)
    {
        int parentVal = desc[0], childVal = desc[1], isLeft = desc[2];
        if (nodes.find(parentVal) == nodes.end())
        {
            nodes[parentVal] = new TreeNode(parentVal);
        }
        if (nodes.find(childVal) == nodes.end())
        {
            nodes[childVal] = new TreeNode(childVal);
        }
        if (isLeft)
        {
            nodes[parentVal]->left = nodes[childVal];
        }
        else
        {
            nodes[parentVal]->right = nodes[childVal];
        }
        hasParent[childVal] = true;
    }

    TreeNode *root = nullptr;
    for (const auto &node : nodes)
    {
        if (hasParent.find(node.first) == hasParent.end())
        {
            root = node.second;
            break;
        }
    }

    return root;
}

int main()
{
    return 0;
}